src/mame/namco/namcos21_3d.cpp

```diff
@@ -84,93 +86,79 @@ void namcos21_3d_device::copy_visible_poly_framebuffer(bitmap_ind16 &bitmap, con
 
 /*********************************************************************************************/
 
-#define SWAP(T,A,B) { const T *temp = A; A = B; B = temp; }
-
-void namcos21_3d_device::renderscanline_flat(const edge *e1, const edge *e2, int sy, unsigned color, int depthcueenable)
+void namcos21_3d_device::renderscanline_flat(const edge *e1, const edge *e2, unsigned sy, unsigned color, bool depthcueenable)
 {
 	if (e1->x > e2->x)
+		std::swap(e1, e2);
+
+	int x0 = (int)e1->x;
+	int x1 = (int)e2->x;
+	int w = x1 - x0;
+
+	if (!w)
+		return;
+
+	double z = e1->z;
+	double dz = (e2->z - e1->z) / w;
+
+	if (x0 < 0)
 	{
-		SWAP(edge, e1, e2);
+		z += -x0 * dz;
+		x0 = 0;
 	}
 
+	if (x1 > m_poly_frame_width - 1)
+		x1 = m_poly_frame_width - 1;
+
+	uint16_t *pDest = m_mpPolyFrameBufferPens.get() + sy * m_poly_frame_width;
+	uint16_t *pZBuf = m_mpPolyFrameBufferZ.get() + sy * m_poly_frame_width;
+
+	for (int x = x0; x < x1; x++)
 	{
-		uint16_t *pDest = m_mpPolyFrameBufferPens.get() + sy * m_poly_frame_width;
-		uint16_t *pZBuf = m_mpPolyFrameBufferZ.get() + sy * m_poly_frame_width;
-		int x0 = (int)e1->x;
-		int x1 = (int)e2->x;
-		int w = x1 - x0;
-		if (w)
+		uint16_t zz = (uint16_t)z;
+
+		if (zz < pZBuf[x])
 		{
-			double z = e1->z;
-			double dz = (e2->z - e1->z) / w;
-			int x, crop;
-			crop = -x0;
-			if (crop > 0)
-			{
-				z += crop * dz;
-				x0 = 0;
-			}
-			if (x1 > m_poly_frame_width - 1)
-			{
-				x1 = m_poly_frame_width - 1;
-			}
+			unsigned pen = color;
 
-			for (x = x0; x < x1; x++)
+			if (depthcueenable && zz > 0)
 			{
-				uint16_t zz = (uint16_t)z;
-				if (zz < pZBuf[x])
-				{
-					int pen = color;
-					if (depthcueenable && zz > 0)
-					{
-						int depth = 0;
-						if (m_depth_reverse)
-						{
-							depth = (zz >> m_zz_shift)*m_zzmult;
-							pen += depth;
-						}
-						else
-						{
-							depth = (zz >> m_zz_shift)*m_zzmult;
-							pen -= depth;
-						}
-					}
-					pDest[x] = pen;
-					pZBuf[x] = zz;
-				}
-				z += dz;
+				const unsigned depth = (zz >> m_zz_shift)*m_zzmult;
+
+				if (m_depth_reverse)
+					pen += depth;
+				else
+					pen -= depth;
 			}
+
+			pDest[x] = pen;
+			pZBuf[x] = zz;
 		}
+
+		z += dz;
 	}
 }
 
-void namcos21_3d_device::rendertri(const n21_vertex *v0, const n21_vertex *v1, const n21_vertex *v2, unsigned color, int depthcueenable)
+void namcos21_3d_device::rendertri(const n21_vertex *v0, const n21_vertex *v1, const n21_vertex *v2, unsigned color, bool depthcueenable)
 {
-	int dy, ystart, yend, crop;
-
 	/* first, sort so that v0->y <= v1->y <= v2->y */
-	for (;;)
 	{
 		if (v0->y > v1->y)
-		{
-			SWAP(n21_vertex, v0, v1);
-		}
-		else if (v1->y > v2->y)
-		{
-			SWAP(n21_vertex, v1, v2);
-		}
-		else
-		{
-			break;
-		}
+			std::swap(v0, v1);
+
+		if (v1->y > v2->y)
+			std::swap(v1, v2);
+
+		if (v0->y > v1->y)
+			std::swap(v0, v1);
 	}
```

Comment:
You’ve broken this – consider what happens if you get (3 2 1).  The first swap makes it (3 1 2), and the second swap makes it (1 3 2).  You need three comparisons to bubble sort three items.

Reply:
Did you really proof read it? There are _three_ comparisons.